### PR TITLE
Proposal: Add `end` tag support

### PR DESCRIPTION
**Pull request by charlespwd**
This PR demonstrates support for the generic and shorter `{% end %}` tag. 

## Behaviour

When used, it will close the closest tag that wasn't closed.

```liquid
{% # still valid %}
{% if true %}
  {% unless false %}
    some stuff
  {% endunless %}
{% endif %}

{% # new! ✨✨ %}
{% if true %}
  {% unless false %}
    yay!
  {% end %}
{% end %}
```

Because of the exceptional behaviour of the `raw` and `comment` tags around nesting, those still require a complete `endraw` and `endcomment` tag to close.

```liquid
{% # still works %}
{% comment %}
  {% end %}
{% endcomment %}

{% # still works %}
{% raw %}
  {% comment %}
    {% end %}
  {% endcomment %}
{% endraw %}

{% # won't works %}
{% raw %}
  {% comment %}
    {% end %}
  {% end %} -- this one would close the raw tag. 
{% end %}

{% comment %}
  In other words, end doesn't close endraw or endcomment
{% endcomment %}
```

## Why make `comment` and `raw` exceptional?

Because you should be able to put a comment anywhere in the code and close it anywhere else. If `end` closes the comment, then that's no longer possible. 

Also, it would be impossible to make this work:
```ruby
assert_template_result(
  '{% comment %} test {% end %}',
  '{% raw %}{% comment %} test {% end %}{% end %}',
)
```

### CODE REVIEW

File: lib/liquid/block_body.rb [hunk 1]
```diff
@@ -68,7 +68,7 @@ def freeze
 
     # @api private
     def self.unknown_tag_in_liquid_tag(tag, parse_context)
-      Block.raise_unknown_tag(tag, 'liquid', '%}', parse_context)
+      Block.raise_unknown_tag(tag, 'liquid', '%}', parse_context, false)
```
[the remark this answers is not in the document]
Comment (charlespwd) [reply lib/liquid/block_body.rb:71]: `raise_unknown_tag` is a class method. We need a way to pass down `supports_end_tag` so that the error message for using an `end` in a `liquid` tag is the following:
```syntax error (line 3): 'end' is not a valid delimiter for liquid tags. use %}```

and not
```syntax error (line 3): 'end' is not a valid delimiter for liquid tags. use end or %}```

(which would be invalid).

File: test/integration/tags/liquid_tag_test.rb [hunk 2]
```diff
@@ -106,6 +106,11 @@ def test_cannot_close_blocks_created_before_a_liquid_tag
       42
       {%- liquid endif -%}
     LIQUID
+    assert_match_syntax_error("syntax error (line 3): 'end' is not a valid delimiter for liquid tags. use %}", <<~LIQUID)
+      {%- if true -%}
+      42
+      {%- liquid end -%}
+    LIQUID
```
[the remark this answers is not in the document]
Comment (charlespwd) [reply test/integration/tags/liquid_tag_test.rb:113]: Good catch! We'll need an exception for that. 

https://github.com/Shopify/theme-tools/issues/479

File: lib/liquid/block_body.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (charlespwd) [reply lib/liquid/block_body.rb:71]: It's not the cleanest code, but I wanted to get alignment first before I spend my time on making this super clean 🙃

**Review by samdoiron — APPROVED** [review]
LGTM from a technical perspective. My only concern was backwards compatibility since we default to supporting the end tag, but AFAIK the usual pattern for breaking the rules around tag scope is to extend `Raw`, which is opted out.

**Comment by jstorie** [comment]
Devils advocate here :) I do understand this would make _writing_ Liquid code a bit simpler, and would match up with how ruby itself works). However, since Liquid is a completely unstructured language that has no whitespace/newline requirements (unlike ruby), this change could reduce readability. In practice, themes can sometimes turn into the wild west with code from multiple sources added with no sensible indentation, and this could make it harder to visually make sense of what's going on. Just something to consider!

**Comment by tyleralsbury** [comment]
> Devils advocate here :) I do understand this would make _writing_ Liquid code a bit simpler, and would match up with how ruby itself works. However, since Liquid is a completely unstructured language that has no whitespace/newline requirements (unlike ruby), this change could reduce readability. In practice, themes can sometimes turn into the wild west with code from multiple sources added with no sensible indentation, and this could make it harder to visually make sense of what's going on. Just something to consider!

I can imagine irresponsible use leading to closures in weird places... but I don't think that's enough to say "we shouldn't do this".  Lots of languages let lots of developers do lots of bad things (including Liquid).  It's up to the theme developer to not do those bad things, in my opinion.

**Comment by isaacbowen** [comment]
*\*thinks about other Liquid implementations, all of which are for host languages other than Ruby*

**Comment by isaacbowen** [comment]
@tobi 👋 ❓

**Comment by mattsodomsky** [comment]
> > My thoughts were similar, and this is a message I wrote in our internal Slack:
> > Yeah, it’s a good fit for a programming language but not a great fit for a templating language is my first response to this. It only started to make sense when I thought about it in the context of Ruby, but that’s not really the context we’re talking about.
> > @charlespwd - can you talk more about the need for this and the impetus for the change?
> > We make apps that have users write Liquid code, and so from that perspective, this will add complexity.
> 
> If you mark all the tags as not supporting the end tag you can effectively opt out of this feature entirely, would that make it more acceptable to see this merged?

Yes, that helps for sure :) but, still, I'd love to take a step back and hear about the why/impetus for this change. It would help my mental model.

**Review by karreiro — APPROVED** [review]
Thank you, @charlespwd!

I personally really like this approach. Most human-friendly languages support the `end` keyword (such as Ruby and Lua).

I believe this helps in making the template a bit cleaner and less verbose. Keeping in mind that we may have nested `if/endif` statements, having the `end` keyword may encourage writers even more to rely on indentation to organize their code.

Still, before merging, as you mentioned, I believe it's worth involving a broader audience, since Liquid has been written differently for years, and we want to make sure that most Liquid users are happy with this change.

Thanks again for this proposal, and for demonstrating its viability! 🔥🚀

**Comment by charlespwd** [comment]
> If concise syntax (shorter, without sacrificing readability) is a goal, then the often requested “inline conditions” (or “ternary expressions” if you prefer) feature would have more of an impact.

This PR is not mutually exclusive to this other proposal ^^ this one is far easier to implement though.

**Comment by benjaminsehl** [comment]
lfg

**Comment by kinngh** [comment]
To add on to this, I want to make this point across in a way where you might see where I think this is a bad idea:
`This is a sentence in the English language. "This is a very good sentence, indeed!", he replied.`

Now, if we follow the exact same rule here, this becomes:
`This is a sentence in the English language. "This is a very good sentence, indeed!., he replied.`

Linting is not going to fix this - This creates more confusion and divides the language, and I hope you can see why this is just a bad idea.

**Comment by panoply** [comment]
I agree with @kinngh.

Sometimes verbosity is a good thing, especially in a template language Like Liquid. There is no added benefit here.

Bad PR overall.

**Comment by jg-rp** [comment]
Note that this PR was based on and merged into the old `master` branch, not the active `main` branch.

**Comment by panoply** [comment]
Introducing an `{% end %}` syntax is akin to suggesting an `</>` as a closing tag in markup. It's about as practical as the subset of HTML implicitly closed tags. Sure, nice to have but necessity is mostly void (no pun intended). 

There is a misalignment here. This new syntactic does not sufficiently account or consider the exhaustive nesting and structural depths inherent with Liquid or how it's implemented by developers. If anything, it completely neglects the appropriated syntactical structures apparent in 90% of Shopify themes. The only outcome here beyond the sugar extrapolation is likely to be the incurred readability issues. The argument for "linting" tools is also assumptive because only a small percent of developers are actually using them, but in any sense they should not be dictating implementation into a language. 

Optionality between the explicit `{% endtag %}` and the implicit `{% end %}` introduces a significant amount of ambiguity about what constitutes correctness. This ambiguity undermines the natural intuitiveness of the language. The rationale behind the decision might make sense when considering it from the viewpoint of serving a community of seasoned developers. However, this approach doesn't align with the reality, as we're discussing a template language primarily used by developers who are often novices, merchants and similar.

It's important to note that Liquid employs keyword control grammar for its execution, rather than relying on character termination (outside of delimiter formation). Therefore, it's crucial to balance this approach, and its imperative to respect consistency as maintainability can be compromised in applied usage.

While I can (personally) reason with this when considering it from the perspective of someone with a more well versed understanding of programming (and to a degree I'd champion something like this in most cases) but I am unconvinced of its merits in practicality for Liquid. I am unable to formulate compelling counter arguments in its favor.

---

~ Original Post on [X](https://x.com/niksavvidis/status/1833687665788764169)

**Comment by TeamDijon** [comment]
Let's go with more RFC 😍
